lxd/api_1.0.go

```diff
@@ -984,6 +985,99 @@ func doAPI10UpdateTriggers(d *Daemon, nodeChanged, clusterChanged map[string]str
 		}
 	}
 
+	value, ok = clusterChanged["instances.migration.stateful"]
```

Comment:
This appears to be now explicitly writing the setting into the profiles DB records rather than server's config?

This is at odds with what we discussed yesterday:

https://github.com/canonical/lxd/pull/12832#discussion_r1485853729

> We agreed to make instances.migration.stateful apply dynamically at start time by default if not overridden in the profile and instance configs. Rather than write a config entry into the new instance DB record.

The only difference now is that you're writing the setting into the profile rather than the instance DB record.

https://github.com/canonical/lxd/pull/12832#issuecomment-1938902897

> We discussed this in the meeting last week, but to clarify, the instances.migration.stateful setting will be dynamically applied to instances at next start time if it hasn't been explicitly overridden in the profile(s) or instance config.

The key word here is "dynamically", the only DB record that should be being updated is the server's global config record, not instances, not profiles.

Then the `migration.stateful` setting will be *dynamically* calculated at the VM's start time based on the server's instances.migration.stateful setting and the profile(s) migration.stateful and the instance's migration.stateful setting, which ever is the lowest level to have it applied (true or false).

Does that make sense?

Reply:
I'll remove this. Sorry again for the misunderstanding. But when you say:

> The key word here is "dynamically", the only DB record that should be being updated is the server's global config record, not instances, not profiles.

The DB record `instances.migration.stateful` is well updated (in `doAPI10Update`) cluster wide, right? So why do we need a particular server handler to set the parameter cluster wide?

As for:

> Then the migration.stateful setting will be dynamically calculated at the VM's start time based on the server's instances.migration.stateful setting and the profile(s) migration.stateful and the instance's migration.stateful setting, which ever is the lowest level to have it applied (true or false).

to get this right once and for all :) does it mean that I can do that? :

```go
// validateStartup checks any constraints that would prevent start up from succeeding under normal circumstances.
func (d *qemu) validateStartup(stateful bool, statusCode api.StatusCode) error {
	err := d.common.validateStartup(stateful, statusCode)
	if err != nil {
		return err
	}

	serverMigrationStateful := d.state.GlobalConfig.InstancesMigrationStateful()
	serverMigrationStatefulStr := "false"
	if serverMigrationStateful {
		serverMigrationStatefulStr = "true"
	}

	if d.expandedConfig["migration.stateful"] == "" { // Not set in instance config nor by applied profiles.
		d.expandedConfig["migration.stateful"] = serverMigrationStatefulStr
	}
	
        // No instance/profile DB records have been changed.
	...
}
```

If this is not related again, then I think I'm lost, sorry.

Reply:
We should materialise the effective instance config as part of the normal `func (d *common) expandConfig() ` call that applies profile(s) config before the instance's config.